### PR TITLE
Ruinian/model simply

### PR DESCRIPTION
**Pull request by ruinianxu**
# Description

This PR is for removing span heads and directly predicting text span for NLU model. The current prediction head consists of language modeling, text span start, text span end and fixed value. Text span start, text span end and fixed value heads
are all removed and corresponding contents are treated as text like language modeling head. 
For example, previously with the query input of ```build a fence in a square shape```, its ground-truth is ```{"dialogue_type": "HUMAN_GIVE_COMMAND", "event_sequence": [{"action_type": "BUILD", "schematic": {"text_span": [0, [2, 6]], "filters": {"where_clause": {"AND": [{"pred_text": "has_name", "obj_text": [0, [2, 2]]}, {"pred_text": "has_tag", "obj_text": [0, [5, 6]]}]}}}}]}``` where integer represents start or end index of query. After change, 
its ground-truth becomes ```{"dialogue_type": "HUMAN_GIVE_COMMAND", "event_sequence": [{"action_type": "BUILD", "schematic": {"text_span": "fence in a square shape", "filters": {"where_clause": {"AND": [{"pred_text": "has_name", "obj_text": "fence"}, {"pred_text": "has_tag", "obj_text": "square shape"}]}}}}]}```. Besides, in ```test_model_script.py```, new feature that evaluating given dataset with beam search is added.

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Previously, prediction heads are composed of language modeling, span begin, span end, text span start, text span end and fixed value. Right now, span begin and span end are removed.

# Testing
First of all, fetch new datasets and model.
To train the model, you could run
```
python droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py --batch_size 8 --param_update_freq 3 --data_dir droidlet/artifacts/datasets/annotated_data/ --dtype_samples 'annotated:1.0' --tree_voc_file droidlet/artifacts/models/nlu/ttad_bert_updated/caip_test_model_tree.json --output_dir ./test_outputs --num_epochs 1
```
To test the performance of trained model, you could run
```
ipython
from droidlet.perception.semantic_parsing.nsp_transformer_model.test_model_script import *
args = argument_parse("")
model, tokenizer = model_configure(args)
dataset = dataset_configure(args, tokenizer)
eval_model(args, model, tokenizer, dataset) or eval_model_bm(args, model, tokenizer, dataset)
```

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

## Performance/Error analysis
On the valid set, it achieves 42.1% accuracy. All results are stored [here](https://docs.google.com/document/d/1Lj5dtqMWzuFWFOBhEfIwLFj7R455E4FTOya2BUhocH4/edit?usp=sharing). Among all 804 failure cases, there are only 4 cases failed due to poorly-formed tree. Among the rest cases, there are 31, 0, 54, 115, 0 and 480 cases failed due to incorrect categorical node key prediction, incorrect fix value node key prediction, incorrect span node key prediction, incorrect categorical node value prediction, incorrect fix value node value prediction and incorrect span node value prediction. Besides the above failure cases, there are 120 cases failed due to the model predict too much or too less of the groud-truth. In other word, the predicted tree is a subset or contains the ground-truth tree. Overall, the major bottleneck for us to conquer is how to predict correct context of span node. All failure cases of span node are stored [here](https://docs.google.com/document/d/1i0xFJsTsZmeb8CtL-an4ASqKNxrx9HMnyZhe6BZasjM/edit?usp=sharing). [Pointer network](https://proceedings.neurips.cc/paper/2015/file/29921001f2f04bd3baee84a12e98098f-Paper.pdf) might be a good direction.

### CODE REVIEW

File: droidlet/tools/artifact_scripts/tracked_checksums/datasets.txt [hunk 1]
```diff
@@ -1 +1 @@
-7421f7de5cabac42fb5a78651d0855fc44516cae
+f0c28e8acad0a327d2431d6acdc7700dd9b5a89c
```
Comment (kavyasrinet) [inline droidlet/tools/artifact_scripts/tracked_checksums/datasets.txt:1]: Is this the same as what's in #1263 ?

**Review by kavyasrinet — COMMENTED** [review]
I'm going to rely on @aszlam 's review on the modeling side given he led the direction of the proposed idea

File: droidlet/tools/artifact_scripts/tracked_checksums/datasets.txt (hunk 1 above)
Comment (ruinianxu) [reply droidlet/tools/artifact_scripts/tracked_checksums/datasets.txt:1]: Yeah, it is. Since i am not sure which PR will be merged first, just in case, I update datasets.txt in both PRs.

File: droidlet/perception/semantic_parsing/nsp_transformer_model/caip_dataset.py [hunk 2]
```diff
@@ -120,30 +120,20 @@ def __getitem__(self, idx):
...
-                ("<S>", -1, -1, -1, -1, -1)
-            ]
-            + tree
-            + [("</S>", -1, -1, -1, -1, -1)]
-        ]
+        tree_idx_ls = self.tokenizer.convert_tokens_to_ids(tree.split())
+
+        # Truncate tokens if it exceeds the maximal length
+        # which is 512 for bert tokenizer
+        max_length = 512
```
Comment (aszlam) [inline droidlet/perception/semantic_parsing/nsp_transformer_model/caip_dataset.py:131]: this needs to come from a config or opt, and be much more visible.  As it is, it will silently truncate

File: droidlet/perception/semantic_parsing/nsp_transformer_model/optimizer_warmup.py [hunk 3]
```diff
@@ -39,9 +49,15 @@ def __init__(self, model, args):
         self._step = 0
 
     def _update_rate(self, stack):
-        return self.lr[stack] * min(
-            (self._step / self.warmup_steps[stack]), (self._step / self.warmup_steps[stack]) ** 0.5
-        )
+        if self._step < self.warmup_steps[stack]:
```
Comment (aszlam) [inline droidlet/perception/semantic_parsing/nsp_transformer_model/optimizer_warmup.py:52]: is this copied from somewhere? can you cite?

File: droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py [hunk 4]
```diff
@@ -537,6 +474,12 @@ def build_grammar(args):
     parser.add_argument(
         "--encoder_learning_rate", default=0.0, type=float, help="Learning rate for the encoder"
     )
+    parser.add_argument(
+        "--encoder_lr_schedules",
```
Comment (aszlam) [inline droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py:478]: can you make a more explicit comment about the algorithm here, and/or cite something if it comes from a paper?

File: droidlet/perception/semantic_parsing/nsp_transformer_model/caip_dataset.py (hunk 2 above)
Comment (ruinianxu) [reply droidlet/perception/semantic_parsing/nsp_transformer_model/caip_dataset.py:131]: > this needs to come from a config or opt, and be much more visible. As it is, it will silently truncate

I tried to retrieve from tokenizer but couldn't find corresponding attribute. I will move it to opt.

File: droidlet/perception/semantic_parsing/nsp_transformer_model/optimizer_warmup.py (hunk 3 above)
Comment (ruinianxu) [reply droidlet/perception/semantic_parsing/nsp_transformer_model/optimizer_warmup.py:52]: This is implemented from [multi_step_lr_scheduler](https://github.com/facebookresearch/mmf/blob/582c7195cbf1eb948436b66c1e9e4bb2e5652a27/mmf/modules/schedulers.py) in [mmf](https://github.com/facebookresearch/mmf). But much simpler. I will improve it to a complete version.

File: droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py (hunk 4 above)
Comment (ruinianxu) [reply droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py:478]: > can you make a more explicit comment about the algorithm here, and/or cite something if it comes from a paper?

Answered above.

**Review by aszlam — APPROVED** [review]
